### ISSUE TITLE
exynos4: trustzone normal side interrupts

### ISSUE THREAD

**Issue by rmillo** [ssumpf/foc#13]
In trustzone normal side execution must not interfere with secure side. Each execution mode has assigned a type of interrupt, the normal side only uses the IRQ and secure side the FIQ.
By default when not using TrustZone Secure Side, Fiasco.OC try to handle (enable/disable) both interrupts, the IRQ and FIQ. We think that there is a problem, when the kernel running on TrustZone Normal Side try to write the FIQ bit, and interrupts remains disabled (the operation is cancelled).

**Comment by rmillo** [comment]
We have adapted the Proc class to activate only the IRQ when compiling FiascoOC on TrustZone Normal Side with a Secure Monitor Interface (85254e4ae070b227badd8f4d80412017808206e7). In this case the FIQ bit is not write and the operation isn't cancelled.
We have tested the patch in Odroid-X2 (Exynos4412) platform, using GenodeOS and L4re.

**Comment by skalk** [comment]
As far as I understood you correctly, you think it is a problem when the OS in the normal world disables FIQs. But that should be no issue at all. The secure world software stack has the responsibility to configure the machine in a way that FIQs are always delivered to the secure world, regardless of the PSR or GIC state of the normal world. On the one hand, when an interrupt is marked as secure at the interrupt controller, the normal world OS cannot change its state, on the other hand one can configure the machine that it traps to the secure world in case of a FIQ regardless of the CPSR.FIQ state of the normal world.

**Comment by rmillo** [comment]
Yes, I think that there is a problem when the OS enables/disables the FIQs in Normal Side or maybe, we are doing something wrong. We are not running nothing in Secure Side, all the system is running in Normal Side.
We think the problem is that the FIQs are marked to be used in Secure Side, in Normal Side it can't  be modified, and Fiasco.OC is trying to modified the FIQ bit when enables the interrupts. We have tested the interrupts in the function `Kernel_thread::bootstrap()` in the file `src/kernel/fiasco/src/kern/kernel_thread.cpp`

``` cpp
Per_cpu_data::run_late_ctors(Cpu::invalid());
bootstrap_arch();

Timer_tick::enable(current_cpu());
Proc::sti();

++ Proc::Status st = Proc::interrupts();
++ if(st)
++    printf("IRQs are enabled\n"); 
```

We have tested it in FiascoOC's implementation, and the IRQs always are disabled and Genode stops the execution. When we have tested it on the modified Fiasco.OC with the proposed patch, the IRQs are enabled and Genode runs succesfully.

**Comment by skalk** [comment]
> We are not running nothing in Secure Side, all the system is running in Normal Side.

Ok, sorry I didn't realized in the first place the actual problem. Thanks for your explanations!
